### ISSUE TITLE
gfx906 DFlash decode: ~17% dispatch overhead in steady-state — three-lever proposal

### ISSUE THREAD

**Issue by unverbraucht** [Kaden-Schutt/hipfire#172]
## Summary

DFlash speculative decode on gfx906 has **~17 % dispatch overhead in
steady-state decode** — diagnostic-driven follow-up to PR #158
(decode optimizations) and `01cc87e` (MMQ cutover fix). Smaller-scope
than the kernel-level wins in PR #158, but addressable with three
focused changes.

(Note: a casual `amdgpu_top` reading shows long stretches of 20-30 %
GFX activity which would suggest 50-70 % idle. That's a window-average
artifact — most of it is the **model-load** copyBuffer chain
preceding decode, not dispatch overhead during decode itself. Filter
to steady-state via `rocprofv3 --kernel-trace` to see the real number.)

## Diagnostic

`rocprofv3 --kernel-trace` filtered to **steady-state decode only**
(last 1.5s of a 58s `dflash_spec_demo` run on Qwen 3.5 27B mq4 + its
drafter, humaneval-0 prompt, `--max 16 --no-chatml`):

```
Steady-state decode: 7768 kernels, wall 1500 ms, GPU busy 1251 ms (83.4%)
Inter-kernel gap distribution (n=1357):
  median: 10.2 us         <- normal HIP launch overhead
  p90   : 505.3 us
  p99   : 1450.6 us
  max   : 12007 us        <- single-shot, cycle boundary
```

Decode dispatch overhead = 250 ms idle / 1500 ms wall = **~17 %**.

By comparison, the pre-decode portion of the same run:
```
Pre-decode (load + prefill): 56500 ms wall, 457 ms GPU-busy (0.8%)
65 gaps > 100 ms, all copyBuffer→copyBuffer (model H2D), total 55.6 s
```

So the production-relevant number is the 17 % steady-state, not the
99.2 % pre-decode (which is one-time load cost).

## Where the dispatch overhead is

The 5 longest decode gaps (~1450-1850 µs each) all share the pattern:

```
after gemm_hfq4g256_residual_mmq_gfx906_full_*  -> before __amd_rocclr_fillBufferAligned
```

`fillBufferAligned` is HIP's internal name for the per-cycle
output-buffer zero memsets (see e.g. `dispatch.rs:7353/7423/7461` for
the residual-GEMM pre-zeros, plus other scratch zeros). They block
the next launch chain by ~1.5 ms each, several times per cycle.

The longest decode gaps (6-12 ms) occur once per spec-cycle at the
boundary between the last verify kernel and the next cycle's first
`embedding_q8` lookup. Cause: per-cycle CPU work (token accept/reject,
kernarg rebuild, kv-cache slot management) + the small sync D2H of
`argmax_buf` (`speculative.rs:2158`, `speculative.rs:2708`).

### 1. Pre-allocate cycle-loop output buffers once, not per-cycle

The per-call `memset_async` of residual-GEMM outputs (e.g. lines
7353/7423/7461 in `dispatch.rs`) happens because the dispatch site
defensively zeros on every call. For the steady-state spec-decode
loop where buffer shapes are identical across cycles, the zeros could
be hoisted out of the per-cycle path (one-time at scratch allocation
or once per shape change). Already async-gated to the active stream,
but still 1-2 ms per cycle on 27B.

Estimated lift: 3-5 % decode tok/s.

### 2. Async D2H of argmax_buf with deferred sync

`speculative.rs:2158` / `:2708` do `gpu.hip.memcpy_dtoh(...)` on the
small `B × 4` argmax buffer. Replacing with `memcpy_dtoh_async` + a
`stream_synchronize` deferred until just before the host_idx read
lets the next launch chain start while the small copy is in flight.

Caveat: host_idx consumption happens immediately after the copy in
current code, so the overlap window is small unless the CPU
bookkeeping (accept/reject, kernarg rebuild) gets reordered to fit
between async D2H and sync. Probably 1-2 % in isolation.

### 3. Move accept/reject decision onto GPU

The accept-or-reject test (`compare drafted[i] == argmax_per_pos[i]`,
return commit count) is purely deterministic from B int32s. A
dedicated `dflash_commit_count` kernel could fold the comparison +
scan onto the GPU, replacing the B×4 D2H with a 4-byte D2H of
`commit_count`. Same shape applies draft-side at line 2708.

Estimated lift: 1-3 %, less compelling than #1 because the D2H is
already small. Worth doing for code-clarity and as prep work if a
future tier wants to graph-capture the whole spec_step.

## What's NOT a lever (ruled out by the trace)

- **Full graph-capture of `spec_step_dflash`** — looked attractive
  before the trace, but if steady-state is already 83 % busy with
  mostly small gaps, even perfect graph capture caps at ~17 % lift.
  The blast radius (KV pointer manipulation, dynamic B,
  `pld_spine` / `cactus_delta` / `repeat_penalty` branches) is large
  for a modest payoff. Reconsider only if a future diagnostic shows
  a different bottleneck.
- **Async per-launch kernarg path** — already in use via
  `launch_maybe_blob` (graph-capture-safe blob path).

## Validation methodology

1. `rocprofv3 --kernel-trace -d /tmp/X` filtered to last 1.5s
   (decode steady-state) — confirm GPU-busy % rises and median gap
   stays ≤10 µs.
2. `benchmarks/scripts/bench_dflash_27b_gfx906.sh` (committed in
   `dd6e410`) for end-to-end DFlash 27B tok/s — pinned harness with
   prompt-md5 + binary-md5 + model-md5 reproducibility metadata.
3. `scripts/coherence-gate-dflash.sh` to confirm no token-attractor
   regression.

## Out of scope

- AR decode (batch=1) doesn't have this problem — covered by
  `HIPFIRE_GRAPH=1` graph capture.
- Prefill doesn't have this problem — single-launch-per-layer-stack.
- Full-logits non-greedy path (`download_f32(&logits_batch)`,
  `B × vocab × 4` bytes per cycle, `speculative.rs:2141`/`:2174`/
  `:2719`) is a much bigger payload with different sync
  characteristics — separate analysis if/when temperature sampling
  becomes the production hot path.

## Related

- PR #158 (gfx906 MMQ kernel redesign + AR decode optimizations)
- Commit `01cc87e` (MMQ min_batch 16 → 8)
- `docs/perf-checkpoints/2026-05-05-gfx906-decode-investigation.md`
  follow-up #7 (DFlash drafter & verify-specific kernels)
- `benchmarks/scripts/bench_dflash_27b_gfx906.sh` (commit `dd6e410`)

**Comment by Kaden-Schutt** [comment]
Closing context (partial): PR #177 merged at 9aabdcf (2026-05-07) addresses one of the three levers in this issue's three-lever proposal: removes 6 diagnostic env::var lookups (HIPFIRE_MMQ_DUMP, _K_FILTER, _CALL_FILTER, _LAYER_FILTER, _TRACE, _DIAG_PASSTHROUGH) from the gfx906 MMQ residual hot path. Per-call cost drops from 5-25us to a single bool load.

The other two levers in the original three-lever proposal (if any) are not addressed by #177. Issue can stay open or be closed depending on whether the residual ~17% steady-state overhead is now within tolerance on real gfx906.

**Comment by unverbraucht** [comment]
## Re-baseline 2026-05-08 — closing as overtaken by events

Re-ran Phase 0 on master (`e1e2bb8`, post-PR #158 + commit `9aabdcf`)
on the same MI50 with `dflash_spec_demo` + Qwen 3.5 27B mq4 + drafter,
`rocprofv3 --kernel-trace` over the last 1.5 s of decode steady-state.

| Metric | Original (this issue) | Today | Δ |
|---|---|---|---|
| GPU-busy % | 83.4 % | **96.7 %** | dispatch overhead 17 % → **3.3 %** |
| Inter-op gap p90 | 505 µs | **13.6 µs** | −97 % |
| Inter-op gap p99 | 1451 µs | **43.8 µs** | −97 % |
| `fillBuffer` ops in steady-state | top-5 | **0** | gone |
| Decode tok/s (27B-3.5 / lru_cache) | ~33 implied | **39.89** | +21 % |

The `fillBufferAligned`-pattern that defined the three-lever proposal
is no longer present on gfx906. PR #158's MMQ kernel redesign moved
the residual GEMMs to `_full_add` / `_full_set` accumulate-into-Y
semantics, eliminating the per-cycle pre-zero memsets — and the
dispatcher is already calling `_full_set` (overwrite-Y) for 272 of
400 MMQ residual calls in steady state, so the lever-1 "stop abusing
the residual kernel" fix shape is partially live already.

### Top-10 longest steady-state gaps today

```
   gap_us  | before                               -> after
  9568.2   | __amd_rocclr_copyBuffer              -> embedding_q8_batched
  8996.0   | __amd_rocclr_copyBuffer              -> embedding_q8_batched
  1361.1   | quantize_q8_1_mmq_ds4                -> gemm_..._full_set_x8
   224.0   | __amd_rocclr_copyBuffer              -> mq_rotate_x
   ... (5 more, all 130–220 µs, all copyBuffer→mq_rotate_x)
```

The two ~9 ms gaps are cycle-boundary CPU stalls (token
accept/reject + KV bookkeeping) — what levers B and C were proposed
to address. Two such gaps in 1500 ms = 18 ms = **1.2 % of wall**.
That's well below gfx906's documented ~2× rocprof-induced noise floor
(`feedback_perf_noise_decode.md`), so neither lever can be measured
reliably on this hardware.

### Verdict per lever

- **Lever 1 (per-cycle memset hoist)** — moot on gfx906 (no memsets
  in the steady-state trace). The original framing applied to gfx11
  WMMA paths anyway; that's a separate question and would need its
  own Phase 0 on a 7900 XTX.
- **Lever 2 (async argmax D2H)** — addresses cycle-boundary stalls
  capped at 1.2 % of wall, below noise. Adversarial review also
  flagged that a naive `memcpy_dtoh_async` swap to a pageable
  `Vec<i32>` is implicitly synchronous per the HIP runtime spec, so
  the diff would need pinned-host plumbing to deliver anything.
- **Lever 3 (GPU-side accept/reject)** — same 1.2 % ceiling, much
  larger diff than lever 2. Not worth it.

### Closing

All three numbered levers in the original proposal are either
structurally addressed (lever 1) or below the noise floor (levers 2
and 3) on gfx906. Closing the issue.

Trace + bench artifacts retained locally at `/tmp/phase0-trace/`
and `/tmp/phase0-bench.log` for reference; full Phase 0 report at
`phase0-gfx906-rebaseline.md` in my working tree.

If gfx11 (7900 XTX) shows the WMMA memsets at
`dispatch.rs:8184/8255/8294` as top-5 gaps in its own Phase 0 trace,
that's a separate issue — I'd file it fresh rather than reopen this
one, since the kernel diff and the validation matrix are different.

— Reviewed and re-baselined by Claude Opus 4.7 (1M ctx)

**Comment by unverbraucht** [comment]
Closed per re-baseline above: dispatch overhead is now 3.3 % (vs 17 % when filed). All three levers are either structurally fixed by PR #158 / commit 9aabdcf or below the gfx906 noise floor.
